pkg/securelaunch/launcher/launcher.go

```diff
@@ -6,43 +6,152 @@
 	"github.com/u-root/u-root/pkg/securelaunch/measurement"
-	"github.com/u-root/u-root/pkg/uio"
+	"github.com/u-root/u-root/pkg/securelaunch/tpm"
 )
 
+// BootEntry holds the names and hashes for a kernel and initrd and the command
+// line to use.
+type BootEntry struct {
+	KernelName  string `json:"kernel name"`
+	KernelHash  string `json:"kernel hash"`
+	KernelBytes []byte
+	InitrdName  string `json:"initrd name"`
+	InitrdHash  string `json:"initrd hash"`
+	InitrdBytes []byte
+	Cmdline     string `json:"cmdline"`
+}
+
 // Launcher describes the "launcher" section of policy file.
 type Launcher struct {
-	Type   string            `json:"type"`
-	Params map[string]string `json:"params"`
+	Type        string               `json:"type"`
+	BootEntries map[string]BootEntry `json:"boot entries"`
 }
 
-// MeasureKernel hashes the kernel and extends the measurement into a TPM PCR.
-func (l *Launcher) MeasureKernel() error {
-	kernel := l.Params["kernel"]
+// bootEntry points to the target BootEntry to use.
+var bootEntry *BootEntry
+
+// readAndVerifyFile reads a file and checks its hash against the provided
+// expected hash. If the hashes match, it returns a byte slice of the file.
+func readAndVerifyFile(fileName string, expectedHash string) ([]byte, error) {
+	slaunch.Debug("readAndVerifyFile: reading file '%s'", fileName)
+
+	// Read the file.
+	fileBytes, err := slaunch.GetFileBytes(fileName)
+	if err != nil {
+		slaunch.Debug("readAndVerifyFile: ERR: could not read file '%s': %v", fileName, err)
+		return nil, fmt.Errorf("could not read file '%s': %w", fileName, err)
+	}
+
+	// Get the file's hash.
+	fileHash := hex.EncodeToString(tpm.HashReader(bytes.NewReader(fileBytes)))
+
+	// Compare hashes.
+	if expectedHash != fileHash {
+		slaunch.Debug("readAndVerifyFile: ERR: file hash (%s) does not match expected (%s)", fileHash, expectedHash)
+		return nil, fmt.Errorf("file hash (%s) does not match expected (%s)", fileHash, expectedHash)
+	}
 
-	if err := measurement.HashFile(kernel); err != nil {
+	return fileBytes, nil
+}
+
+// measureFile extends the measurement of the provided file into the TPM.
+func measureFile(fileName string, fileBytes []byte) error {
+	if len(fileBytes) == 0 {
+		return fmt.Errorf("file not yet loaded or empty")
+	}
+
+	eventDesc := fmt.Sprintf("File Collector: measured %s", fileName)
+	if err := measurement.HashBytes(fileBytes, eventDesc); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// MeasureKernel hashes the kernel and extends the measurement into a TPM PCR.
+func MeasureKernel() error {
+	if bootEntry == nil {
+		return fmt.Errorf("boot entry not yet selected")
+	}
+
+	return measureFile(bootEntry.KernelName, bootEntry.KernelBytes)
+}
+
 // MeasureInitrd hashes the initrd and extends the measurement into a TPM PCR.
-func (l *Launcher) MeasureInitrd() error {
-	initrd := l.Params["initrd"]
+func MeasureInitrd() error {
+	if bootEntry == nil {
+		return fmt.Errorf("boot entry not yet selected")
```

Comment:
yeah you def want an error variable, this string appears twice.
var ErrBootNotSelected=errors.New("boot entry not yet selected")

it's a pain at first but once you start writing tests you will be happy.